flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.contrib.streaming.state;
+
+import org.apache.flink.core.fs.CloseableRegistry;
+import org.apache.flink.core.fs.FSDataInputStream;
+import org.apache.flink.core.fs.FSDataOutputStream;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.state.IncrementalKeyedStateHandle;
+import org.apache.flink.runtime.state.StateHandleID;
+import org.apache.flink.runtime.state.StreamStateHandle;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Data transfer utils for RocksDbKeyedStateBackend.
+ */
+public class RocksDbStateDataTransfer {
+
+	public static void transferAllStateDataToDirectory(
+		IncrementalKeyedStateHandle restoreStateHandle,
+		Path dest,
+		int resotreThreadNum,
+		CloseableRegistry closeableRegistry) throws Exception {
+
+		final Map<StateHandleID, StreamStateHandle> sstFiles =
+			restoreStateHandle.getSharedState();
+		final Map<StateHandleID, StreamStateHandle> miscFiles =
+			restoreStateHandle.getPrivateState();
+
+		transferAllDataFromStateHandles(sstFiles, dest, resotreThreadNum, closeableRegistry);
+		transferAllDataFromStateHandles(miscFiles, dest, resotreThreadNum, closeableRegistry);
+	}
+
+	/**
+	 * Copies all the files from the given stream state handles to the given path, renaming the files w.r.t. their
+	 * {@link StateHandleID}.
+	 */
+	private static void transferAllDataFromStateHandles(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		int resotreThreadNum,
+		CloseableRegistry closeableRegistry
+		) throws Exception {
+
+		final ExecutorService executorService = Executors.newFixedThreadPool(resotreThreadNum);
+		List<CompletableFuture<Void>> futures = new LinkedList<>();
+		List<Closeable> closeables = new LinkedList<>();
+
+		try {
+			closeables.add(() -> executorService.shutdownNow());
```

Comment:
I think the tasks should be first properly canceled and then `executorService` should shutdown.
`executorService.shutdownNow` also calls `future.cancel` which is interrupting the task thread.
At the moment `while (true)` in `copyStateDataHandleData` does not support explicit cancelation.
I suggest to create an `running = AtomicBoolean(true)` in the beginning then `while (true)` can be `while(running.get())`. This will allow to cancel all tasks at once and shutdown `executorService` in one `closeable`:
`closeables.add(() -> { running.set(false); executorService.shutdownNow()); };`
List of `closeables` is not needed then.